docs/madr/003-arns-undername-limits.md

```diff
@@ -0,0 +1,237 @@
+# ArNS Undername Limit Enforcement
+
+- Status: proposed
+- Deciders: [Ariel], [Dylan], [Atticus], [Phil], [David]
+- Date: 2025-01-14
+- Authors: [Dylan]
+
+## Context and Problem Statement
+
+ArNS names have a supported undername limit, defined by the ARIO network contract. Increasing this limit requires payment in $ARIO tokens to compensate gateway operators for the additional computational resources needed to serve undername and promote responsible usage of ArNS. AR.IO gateways must enforce this limit when resolving ArNS names to ensure operators are fairly rewarded for their services through fees paid by increasing a name's undername limit and providing a consistent experience across the network.
+
+Currently, the `getRecords` API on ANTs return **a table of records** allowing for efficient undername lookups. Lua does not guarantee the order of keys in a table, making it difficult to know what undernames to enforce. Sorting needs to either 1. be done by the ANT and provided as an array of records, or 2. be done by the [ar-io-node] using additional information provided by the ANT. Additionally, caching within the ar-io-node needs to properly handle these priority order changes.
+
+## Decision Outcome
+
+### Priority Resolution Flow
+
+The following sequence diagram is used to demonstrate how AR.IO gateways resolve ArNS names and enforce undername limits.
+
+```mermaid
+sequenceDiagram
+    participant Client
+    participant AR.IO Gateway
+    participant ARIOContract
+    participant ANTContract
+
+    Client->>AR.IO Gateway: ArNS Name Request
+    AR.IO Gateway->>ANTContract: Check ANT records with priority
+    ANTContract->>AR.IO Gateway: Return records with priority data
+    AR.IO Gateway->>ARIOContract: Get undername limit
+    ARIOContract->>AR.IO Gateway: Return undername limit
+    note over AR.IO Gateway: Sort records by ANT priority, fallback to alphabetical
+    note over AR.IO Gateway: Validate undernames against undername limit
+    alt Undername limit exceeded
+        AR.IO Gateway->>Client: Return 402 (Payment required)
+    else Undername limit not exceeded
+        AR.IO Gateway->>Client: Resolve undername
+    end
+```
+
+## Decision Drivers
+
+- Honor ANT priority ordering when available
+- Consistent enforcement of undername limits across AR.IO gateways
+- Simple fallback mechanism if ANT does not return or contain priority data
```

Comment:
This is a wacky scenario, but what if an ANT doesn't consistently return its map and also doesn't provide prioritization ordering such that different gateways might end up with different valid undernames? Obviously this primarily hurts the users of the undernames, but I'm trying to decide whether this negatively impacts gateways in a way that hurts the protocol. Since the observers don't yet check undernames I think we'd be ok?

Reply:
I think that's something worth pointing out in this ADR: ANTs should be considered untrustworthy 🤔

Reply:
good call out - will add

Reply:
[90be780](https://github.com/ar-io/ar-io-node/pull/272/commits/90be780969e021f3f9cfee168c331a143129ec1e)